src/interfaces/cloud.rs

```diff
@@ -0,0 +1,410 @@
+/// 2. base64 representation of the different keys:
+///     1. `SIGNATURE_KEY_LENGTH` bytes of findex master key (this key is never
+/// sent to the Findex Cloud backend)
+///     2. 1 byte prefix identifying the next key
+///     3. `SIGNATURE_KEY_LENGTH` bytes of callback signature key
+///     4. 1 byte prefix identifying the next key
+///     5. …
+///
+/// Currently each callback have an associated signature key used in a kmac to
+/// send request to the backend. These key are only used for authorization
+/// and do not secure the index (the findex master key do). In the future, we
+/// could do optimization to avoid having one key for each callback but we want
+/// to disallow the server to differentiate a `fetch_entries` for a search or a
+/// `fetch_entries` for an upsert while still allowing fine grain permissions.
+pub(crate) struct Token {
+    /// This ID identify an index inside the Findex Cloud backend
+    /// This number is not sensitive, it's only an ID. If someone find this ID,
+    /// it cannot do requests on the index because it doesn't have the keys.
+    /// We do not use auto-increment integer ID because we don't want to leak
+    /// the number of indexes inside our database.
+    /// We do not use UUID because the token is limited in space.
+    /// The abritrary chosen length is `INDEX_ID_LENGTH`.
+    index_id: String,
+
+    pub(crate) findex_master_key: KeyingMaterial<MASTER_KEY_LENGTH>,
+
+    fetch_entries_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    fetch_chains_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    upsert_entries_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    insert_chains_key: Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
+
+        if let Some(fetch_entries_key) = &self.fetch_entries_key {
+            keys.push(0);
+            keys.extend(fetch_entries_key.as_ref());
+        }
+
+        if let Some(fetch_chains_key) = &self.fetch_chains_key {
+            keys.push(1);
+            keys.extend(fetch_chains_key.as_ref());
+        }
+
+        if let Some(upsert_entries_key) = &self.upsert_entries_key {
+            keys.push(2);
+            keys.extend(upsert_entries_key.as_ref());
+        }
+
+        if let Some(insert_chains_key) = &self.insert_chains_key {
+            keys.push(3);
+            keys.extend(insert_chains_key.as_ref());
+        }
+
+        write!(f, "{}{}", self.index_id, base64::encode(keys))
+    }
+}
+
+impl FromStr for Token {
+    type Err = FindexErr;
+
+    fn from_str(token: &str) -> Result<Self, Self::Err> {
+        let (index_id, tail) = token.split_at(INDEX_ID_LENGTH);
+        let mut bytes = base64::decode(tail)
+            .map_err(|_| {
+                FindexErr::Other(format!(
+                    "token {token} is not a valid base64 encoded string"
```

Comment:
I am a bit skeptical about this assertion since I am not sure the token *is* a valid base64 string. Only `tail` should be. I am pretty sure token is actually *not* a valid base64 string since `INDEX_ID_LENGTH` is 5 which makes the token length an odd number.

Reply:
Changed the wording in 424a022, is it ok?

Reply:
Yes it is better, maybe just change `the keys section are` into `the key section is`?